### PR TITLE
Infra Integration Test v2 with Generate Pipeline

### LINKED ISSUE #872: SPK integration test should validate infra generation pipeline
**As a:**
Developer

**I want:**
reliable integration testing for SPK Infra Generation Pipeline

**So that:**
I can ensure that my Generation pipeline will run smoothly in production.

**Describe the solution you'd like:**
Expand on the current `spk infra` integration tests in `infra-validations.sh` and `functions.sh` to incorporate the following:
- [x] Create and verify an Infra HLD Repository
- [x] Create Variable Group and populate variables needed for generate pipeline
- [x] Download the template `azure-pipelines.yaml` to this repo
- [x] Creation of hierarchy of directories and definitions:
```
|- discovery-service
    |- definition.yaml
    |- west/
        |- definition.yaml
```
- [x] Create generation pipeline (from infra HLD repo)
- [x] Create and verify a Generated Repository
- [x] Verify that generation pipeline runs to success

**Acceptance Criteria:**
Will use [`validations.sh`](https://github.com/CatalystCode/spk/blob/master/tests/validations.sh) as a reference.

**Additional context:**
This task should be performed in alongside https://github.com/microsoft/bedrock/issues/871.

**Does this require updates to documentation?:**
Yes

### PR DESCRIPTION
**Pull request by NathanielRose**
Closes [#872](https://github.com/microsoft/bedrock/issues/872)
- Added revised integration test for Infra deployment
- Validates `spk infra scaffold` `spk infra generate`
- Requires `SUBSCRIPTION_ID` be added to `spk-vg`
- Requires `INFRA_YML` be added to `spk-vg`

SmokeTest Validation (Failed due to pending queued pipe for subscription) - https://naros32.visualstudio.com/spk-testing/_build/results?buildId=275&view=results
Pipeline Validation - https://naros32.visualstudio.com/spk-testing/_build?definitionId=90&_a=summary

### CODE REVIEW

File: azure-pipelines/templates/infra-generation-pipeline.yml [hunk 1]
```diff
@@ -115,12 +120,16 @@ steps:
     repo_url="${repo_url#https://}"
     echo "$repo_url"
     echo "GIT PUSH: https://<ACCESS_TOKEN_SECRET>@$repo_url"
+    echo $repo_url
+    echo $ACCESS_TOKEN_SECRET
```
Comment (yradsmikham) [inline azure-pipelines/templates/infra-generation-pipeline.yml]: We probably should not echo this.

File: tests/infra-validations.sh [hunk 2]
```diff
@@ -78,41 +85,145 @@ echo "The remote_repo_url is $remote_repo_url"
...
+
 # Single Cluster scaffold for template
 pwd
 echo "../$terraform_template_dir"
 echo "$tf_template_version"
-spk infra scaffold -n $infra_hld_dir --source "$source" --version "$tf_template_version" --template "template" >> $TEST_WORKSPACE/log.txt
+mkdir $infra_hld_dir
+cd $infra_hld_dir
+
+spk infra scaffold -n $infra_hld_project --source "$source" --version "$tf_template_version" --template "template" >> $TEST_WORKSPACE/log.txt
```
Comment (yradsmikham) [inline tests/infra-validations.sh:108]: How can users view `spk infra scaffold` logs?

**Review by yradsmikham — CHANGES_REQUESTED** [review]

File: azure-pipelines/templates/infra-generation-pipeline.yml (hunk 1 above)
Comment (NathanielRose) [reply azure-pipelines/templates/infra-generation-pipeline.yml]: Removed.

File: tests/infra-validations.sh (hunk 2 above)
Comment (NathanielRose) [reply tests/infra-validations.sh:108]: Logs are published as artifacts in the smoke test pipeline. See: https://github.com/CatalystCode/spk/blob/master/smoke-test-pipeline.yml#L79

File: tests/infra-validations.sh [hunk 3]
```diff
@@ -78,41 +85,145 @@ echo "The remote_repo_url is $remote_repo_url"
...
+# Create the remote infra hld repo for the local repo
+created_repo_result=$(az repos create --name "$infra_hld_dir" --org $AZDO_ORG_URL --p $AZDO_PROJECT) >> $TEST_WORKSPACE/log.txt
+
+# Extract out remote repo URL from the above result
+remote_repo_url=$(echo $created_repo_result | jq '.remoteUrl' | tr -d '"' )
+echo "The remote_repo_url is $remote_repo_url"
+
+# Remove the user from the URL
+repo_url=$(getHostandPath "$remote_repo_url")
+git commit -m "inital commit"
```
Comment (yradsmikham) [inline tests/infra-validations.sh]: Should we maybe have a more meaningful commit message here?

File: tests/infra-validations.sh [hunk 4]
```diff
@@ -78,41 +85,145 @@ echo "The remote_repo_url is $remote_repo_url"
...
+# Does variable group already exist? Delete if so
+variable_group_exists $AZDO_ORG_URL $AZDO_PROJECT $vg_name "delete"
+
+# Create variable group
+az pipelines variable-group create --name $vg_name --authorize true --variables "ACCESS_TOKEN_SECRET=$ACCESS_TOKEN_SECRET" "ARM_CLIENT_ID=$SP_APP_ID" "ARM_CLIENT_SECRET=$SP_PASS" "ARM_SUBSCRIPTION_ID=Tentative" "ARM_TENANT_ID=$SP_TENANT" "CLUSTER=$infra_region" "GENERATED_REPO=https://$repo_url" "PROJECT_DIRECTORY=$infra_hld_project" "AZDO_ORG_NAME=$AZDO_ORG_URL" "AZDO_PROJECT_NAME=$AZDO_PROJECT" "ARM_SUBSCRIPTION_ID=$SP_SUBSCRIPTION_ID"
+
+# Verify the variable group was created. Fail if not
+variable_group_exists $AZDO_ORG_URL $AZDO_PROJECT $vg_name "fail"
+
+
```
Comment (yradsmikham) [inline tests/infra-validations.sh]: remove empty line.

**Review by yradsmikham — CHANGES_REQUESTED** [review]

**Review by yradsmikham — APPROVED** [review]

**Review by andrebriggs — APPROVED** [review]
@NathanielRose looks good. Do we have list of next steps we want to test in this area?

**Comment by NathanielRose** [comment]
> @NathanielRose looks good. Do we have list of next steps we want to test in this area?

@andrebriggs Next steps items are listed here: https://github.com/microsoft/bedrock/issues/919

File: smoke-test-pipeline.yml [hunk 5]
```diff
@@ -37,10 +37,18 @@ stages:
       failOnStderr: true
     - bash: |
         echo "FUNC_SCRIPT: $FUNC_SCRIPT"
-        curl $FUNC_SCRIPT >| functions.sh
```
Comment (andrebriggs) [inline smoke-test-pipeline.yml:40]: @NathanielRose here

File: smoke-test-pipeline.yml [hunk 6]
```diff
@@ -37,10 +37,18 @@ stages:
...
+        curl --fail $FUNC_SCRIPT > function.sh 2>/dev/null
+        if [ $? -ne 0 ]; then
+            echo "Error finding FUNC_SCRIPT: $FUNC_SCRIPT"
+            exit 1
+        fi
         chmod +x ./functions.sh
         echo "TEST_SCRIPT: $TEST_SCRIPT"
-        curl $TEST_SCRIPT >| validations.sh
```
Comment (andrebriggs) [inline smoke-test-pipeline.yml:43]: @NathanielRose and here

File: smoke-test-pipeline.yml [hunk 7]
```diff
@@ -105,10 +113,25 @@ stages:
       failOnStderr: true
     - bash: |
         echo "FUNC_SCRIPT: $FUNC_SCRIPT"
-        curl $FUNC_SCRIPT >| functions.sh
```
Comment (andrebriggs) [inline smoke-test-pipeline.yml:108]: @NathanielRose

**Pull request merged**